backend/app/notify/email.go

```diff
@@ -0,0 +1,380 @@
+	if res.TimeOut <= 0 {
+		res.TimeOut = defaultEmailTimeout
+	}
+	if res.FlushDuration <= 0 {
+		res.FlushDuration = defaultFlushDuration
+	}
+	if res.BufferSize <= 0 {
+		res.BufferSize = 1
+	}
+	if res.MsgTemplate == "" {
+		res.MsgTemplate = defaultEmailTemplate
+	}
+	if res.VerificationTemplate == "" {
+		res.VerificationTemplate = defaultEmailVerificationTemplate
+	}
+	// unbuffered send channel for sending messages to autoFlush goroutine
+	res.submit = make(chan emailMessage)
+
+	log.Printf("[DEBUG] create new email notifier for server %s with user %s, timeout=%s",
+		res.Host, res.Username, res.TimeOut)
+
+	// initialise templates
+	res.msgTmpl, err = template.New("messageFromRequest").Parse(res.MsgTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse message template")
+	}
+	res.verifyTmpl, err = template.New("messageFromRequest").Parse(res.VerificationTemplate)
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't parse verification template")
+	}
+
+	// establish test connection
+	testSMTPClient, err := res.client()
+	if err != nil {
+		return &res, errors.Wrapf(err, "can't establish test connection")
+	}
+	if err = testSMTPClient.Quit(); err != nil {
+		log.Printf("[WARN] failed to send quit command to %s:%d, %v", res.Host, res.Port, err)
+	}
+	if err = testSMTPClient.Close(); err != nil {
+		log.Printf("[WARN] can't close smtp connection, %v", err)
+	}
+	return &res, err
+}
+
+// Send email from request to request.parentUserEmail if it's set, otherwise do nothing and return nil, thread safe
+// do not returns sending error, except following cases:
+// 1. (likely impossible) template execution error from email message creation from request
+// 2. message dropped without sending in case of closed ctx
+func (e *Email) Send(ctx context.Context, req request) error {
+	if req.parentUserEmail == "" || req.parent.User == req.comment.User {
+		// don't send anything if there is no email to send information to
+		// or if user replied to his own comment
+		return nil
+	}
+	// start auto flush once, as this is the first moment we see the context from caller
+	e.once.Do(func() {
+		// e.smtpClient initialised only in tests
+		go e.autoFlush(ctx, e.smtpClient)
+	})
+	log.Printf("[DEBUG] send notification via %s, comment id %s", e, req.comment.ID)
+	msg, err := e.buildMessageFromRequest(req, req.parentUserEmail)
+	if err != nil {
+		return err
+	}
+	select {
+	case e.submit <- emailMessage{msg, req.parentUserEmail}:
+		return nil
+	case <-ctx.Done():
+		return errors.Errorf("sending message to %q aborted due to canceled context", req.parentUserEmail)
+	}
+}
+
+// SendVerification sends email verification
+// do not returns sending error, except following cases:
+// 1. (likely impossible) template execution error from verify message template execution
+// 2. message dropped without sending in case of closed ctx
+func (e *Email) SendVerification(ctx context.Context, req VerificationRequest) error {
```

Comment:
again, mixing resposibilty. SendVerification is a high-level concern. Having it next to the autoFlush make me concerend a little bit.